File: lidar_pkg/src/lidar_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>
#include <geometry_msgs/msg/twist.hpp>

rclcpp::Node::SharedPtr node;
rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr vel_pub;
int rotate_cycle = 0;

void lidar_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
{
    float front_distance = msg->ranges[180];
    RCLCPP_INFO(node->get_logger(), "The front distance is: %f meters", front_distance);

    geometry_msgs::msg::Twist vel_msg;

    if (rotate_cycle > 0)
    {
        rotate_cycle--;
        return;
    }

    // too close will collide, turn left
    if (front_distance < 1)
    {
        // give the command to turn left
        vel_msg.angular.z = 2.0;
        rotate_cycle = 9;
    }
    else{
        // give the command to go forward
        vel_msg.linear.x = 1.0;
    }

    vel_pub->publish(vel_msg);
}

int main(int argc, char *argv[])
{
    // init rclcpp
    rclcpp::init(argc, argv);

    // print hello and introduction
    printf("Hello World, I am the lidar node!\n");

    // create node
    node = rclcpp::Node::make_shared("lidar_node");

    // create publisher
    vel_pub = node->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);

    // create subscriber
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr sub = node->create_subscription<sensor_msgs::msg::LaserScan>(
        "/scan",
        10,
        lidar_callback
    );

    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}

File: lidar_pkg/lidar_pkg/lidar_node.py
#!/usr/bin/env python3
#conding: utf-8

import rclpy
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist

node = None
vel_pub = None
rotate_cycle = 0

def lidar_callback(msg):
    global node
    global vel_pub
    global rotate_cycle

    front_distance = msg.ranges[180]
    node.get_logger().info(f"The front distance is: {front_distance} meters")

    vel_msg = Twist()

    if rotate_cycle > 0:
        rotate_cycle -= 1
        return

    if front_distance < 1.5:
        vel_msg.angular.z = 2.0
        rotate_cycle = 9
    else:
        vel_msg.linear.x = 2.0

    vel_pub.publish(vel_msg)

def main(args=None):
    global node
    global vel_pub

    rclpy.init(args=None)

    node = rclpy.create_node('lidar_node')

    print("Hello World, I am the lidar node! [Python]\n")

    vel_pub = node.create_publisher(Twist, '/cmd_vel', 10)
    sub = node.create_subscription(LaserScan, '/scan', lidar_callback, 10)

    rclpy.spin(node)

    node.destroy_node()

    rclpy.shutdown()

if __name__ == '__main__':
    main()
